### ISSUE TITLE
update_cognitive_models.ps1 messes up emojis

### ISSUE THREAD

**Issue by Avalaxy** [microsoft/botframework-solutions#2224]
#### What project is affected?
The Virtual Assistant template

#### What language is this in?
C#

#### What happens?
When I run update_cognitive_models.ps1, all the emojis in my QnAmaker answers are changed into unreadable symbols (encoding issue).

#### What are the steps to reproduce this issue?
Add some emojis in qnamaker, run update_cognitive_models.ps1 and see how the emojis are all messed up.

#### What were you expecting to happen?
That the emojis would not be messed up.

#### Can you share any logs, error output, etc.?
For example 🤝 turns into ≡ƒñ¥

**Comment by lauren-mills** [comment]
Hi @Avalaxy, you're right, this is likely an encoding issue. I've added it to our backlog!

**Comment by Avalaxy** [comment]
Is there a workaround for this?

**Comment by lauren-mills** [comment]
I don't have one right now, because I'm not sure if its an issue with the qnamaker CLI tool or the update script itself. I'm adding our support team to help identify the root of the issue.

**Comment by lauren-mills** [comment]
You could try replacing the literal emoji with its unicode equivalent and see if that helps.

**Comment by dmvtech** [comment]
hi @avalaxy Can you give me some more info on how to reproduce this? I don't seem to be able to do so. Are you seeing the emoji change to ≡ƒñ¥ in qnamaker.ai? Can you also share the exact command (with parameters) you are using to call update_cognitive_models.ps1? What _channels_ are you using?

Here is what I did:

- Create a QnA pair in my kb
- Question: Can we shake hands?
- Answer: 🤝
- Test in qnamaker.ai test panel. Success.
- Test in Test in Web Chat, Teams, Skype. All fail with **Sorry, it looks like something went wrong.**
- Run `.\Deployment\Scripts\update_cognitive_models.ps1 -RemoteToLocal`
- Check in qnamaker.ai, still shows fine.

I think I need to be able to reproduce the issue correctly before I can start to isolate and look for work arounds.

**Comment by Avalaxy** [comment]
Hi @dmvtech. The emojis did get changed back in QnAmaker. First I ran `update_cognitive_models.ps1 -RemoteToLocal`, and then I ran `update_cognitive_models.ps1`. I'm not entirely sure what this script does, but it seems like it pulled all the questions from QnAmaker, updated them locally and then wrote them back to QnAmaker? Actually this last step (writing back to QnAmaker) is completely surprising to me, afaik only the LUIS Dispatcher would need to be updated? At least I think this is what was happening in the old luis dispatch refresh command (from the Enterprise Bot Template).

So to resume, the emojis started out fine, but now they are messed up both locally and in the remote QnAmaker. I have to import a backed up version to restore the content to the correct state.

Oh, and the emojis I use are just the unicode emojis, I inspected them. For example 🙂 is 1F642. We use Facebook as a channel, but I don't think that's related because it's going wrong in the dispatcher sync.

**Comment by dmvtech** [comment]
I'm not able to reproduce this. I'm not sure what I'm missing. 

The reason I was asking about the channel, was so that I could get the emojis to show (as I said, they weren't for me in Teams/Skype) and so that I could make sure I had the same repro that you are seeing.

I'm still working on it. Let you know what I find.

**Comment by dmvtech** [comment]
@Avalaxy I'm prety sure this is related to the version of [ludown](https://github.com/microsoft/botbuilder-tools/tree/master/packages/Ludown). 

I was able to reproduce the issue with version `1.3.1`, but works fine with `1.3.3`. I'm really confused on why I have two different versions installed, but that's my own problem at this point 😆

Can you confirm what version of ludown you have (`ludown -v`)?

If running older than `1.3.3`, try running `npm i -g ludown` and then try again.

**Comment by Avalaxy** [comment]
Hi Dana, great to hear that you are able to reproduce it! I should be
running 1.3.4-preview-190803-01. Not 100% sure I ran it with this version
so I will try it when I have time (although I can't remember upgrading
ludown).

Should I run it with -RemoteToLocal or without? Or both?

Op vr 13 sep. 2019 om 20:53 schreef Dana V <notifications@github.com>:

**Comment by dmvtech** [comment]
The issue (for me at least) happens with the following command:
`ludown parse toqna --in <Path> --out_folder <Path> --out <Filename>`

Which happens in the script block when -RemoteToLocal switch is not supplied. Whether you should be [using that switch](https://microsoft.github.io/botframework-solutions/tutorials/csharp/customize-assistant/4_edit_your_cognitive_models/) depends on what you have changed, where, and what you are wanting to accomplish. If you are making changes in the LUIS/QnA portal, then use -RemoteToLocal. If you are making changes to local files (exported, etc) to push to the cloud KBs, then you would not supply that switch. 

Here's essentially what the script does (LUIS or Qna)
```
if($RemoteToLocal)
    {
        # Update local LU files based on hosted QnA KBs
    }
    else
    {
        # Update each knowledgebase based on local LU files
```

**Comment by dmvtech** [comment]
@Avalaxy Just checking in. Any progress? Still having the issue?

**Comment by dmvtech** [comment]
I'm closing this as I think the issue has been identified and the fix is available. If you still continue to have this issue using a newer version of ludown (>1.3.3), let me know and we will assist.

**Comment by Avalaxy** [comment]
@dmvtech nope, it's still not fixed. Just updated ludown to version 1.4, ran -RemoteToLocal to update my local files and then ran it without -RemoteToLocal, and all my emojis are converted to gibberish again. Also, when I run the script I get a message that there is a new CLI tool? What is that now, _again_ a new way of deploying things?

**Comment by CoHealer** [comment]
@lauren-mills, is there a time frame to fix this? Given it's a bug without a work around, there's not a lot more folks on the thread can do at this point.

**Comment by CoHealer** [comment]
@lauren-mills, is there anything that @Avalaxy can do here?

**Comment by darrenj** [comment]
@Avalaxy the BF CLI is a "port" of the qnamker and other CLI tools into one so we reduce the number of CLI tools you need to import.

This seems to be a CLI issue as all the script does is invoke the CLI to perform the conversion. @Avalaxy  to help us out could you provide the text you have in a QnAMaker answer that demonstrates the issue and I can run the appropriate CLI tools independent of the update cognitive models script and then if confirmed can transfer this to the tools team to fix.

**Comment by darrenj** [comment]
Please re-open if you find the issue with the latest bf tool.

**Comment by Avalaxy** [comment]
I did another run of update_cognitive_models.ps1, with the latest version of the file downloaded from here: https://microsoft.github.io/botframework-solutions/virtual-assistant/handbook/deployment-scripts/

Again, it creates a local file where the emojis are broken. I have all the tools updated to the latest version, so I don't know what's up?

Also, is it correct that it now creates a .json file for my content instead of a .qna file?

Anyway, the original text I have is:

A single ticket gives you access to both the permanent collections and temporary exhibition. 🙂

It is then transformed into:

A single ticket gives you access to both the permanent collections and temporary exhibition. ≡ƒÖé

**Comment by GitDruid** [comment]
Hi @Avalaxy ,
try downloading the latest version of the script from the repository.
There was [a fix about encoding issues ](https://github.com/microsoft/botframework-solutions/pull/3557/commits/3955f81374c871d1ac6f4ac6c7289c5144015098)that solved me a problem similar to this #3465.

Bye,
Alessandro.
